azkaban-common/src/main/java/azkaban/executor/AlerterHolder.java

```diff
@@ -16,119 +16,118 @@
+    final File[] pluginDirs = pluginsRootPath.listFiles();
 
     for (final File pluginDir : pluginDirs) {
-      // load plugin properties
-      final Props pluginProps = PropsUtils.loadPluginProps(pluginDir);
-      if (pluginProps == null) {
-        continue;
-      }
-
-      final String pluginName = pluginProps.getString("alerter.name");
-      final List<String> extLibClassPaths =
-          pluginProps.getStringList("alerter.external.classpaths",
-              (List<String>) null);
-
-      final String pluginClass = pluginProps.getString("alerter.class");
-      if (pluginClass == null) {
-        logger.error("Alerter class is not set.");
-        continue;
-      } else {
-        logger.info("Plugin class " + pluginClass);
+      try {
+        loadCustomAlerter(installedAlerterPlugins, parentLoader, pluginDir);
+      } catch (final Exception e) {
+        logger.error(String.format("Failed to load alerter plugin in '%s'.", pluginDir), e);
       }
+    }
+    return installedAlerterPlugins;
+  }
 
-      Class<?> alerterClass =
-          PluginUtils.getPluginClass(pluginClass, pluginDir, extLibClassPaths, parentLoader);
+  private void loadCustomAlerter(final Map<String, Alerter> installedAlerterPlugins,
+      final ClassLoader parentLoader, final File pluginDir) {
+    // Load plugin properties.
+    final Props pluginProps = PropsUtils.loadPluginProps(pluginDir);
+    if (pluginProps == null) {
+      throw new AzkabanException("Plugin config properties could not be loaded.");
+    }
 
-      if (alerterClass == null) {
-        continue;
-      }
+    final String pluginName = pluginProps.getString("alerter.name", "").trim();
+    if (pluginName.isEmpty()) {
+      throw new AzkabanException("Alerter name is required.");
+    }
 
-      final String source = FileIOUtils.getSourcePathFromClass(alerterClass);
-      logger.info("Source jar " + source);
-      jarPaths.add("jar:file:" + source);
+    final List<String> extLibClassPaths = pluginProps.getStringList(
+        "alerter.external.classpaths", (List<String>) null);
 
-      Constructor<?> constructor = null;
-      try {
-        constructor = alerterClass.getConstructor(Props.class);
-      } catch (final NoSuchMethodException e) {
-        logger.error("Constructor not found in " + pluginClass);
-        continue;
-      }
+    final String pluginClass = pluginProps.getString("alerter.class", "").trim();
+    if (pluginClass.isEmpty()) {
+      throw new AzkabanException("Alerter class is required.");
+    }
 
-      Object obj = null;
-      try {
-        obj = constructor.newInstance(pluginProps);
-      } catch (final Exception e) {
-        logger.error(e);
-      }
+    final Class<?> alerterClass =
+        PluginUtils.getPluginClass(pluginClass, pluginDir, extLibClassPaths, parentLoader);
+    if (alerterClass == null) {
+      throw new AzkabanException(
+          String.format("Alerter class '%s' could not be loaded.", pluginClass));
+    }
```

Comment:
old logic will `continue` and here it's breaking, is this still good?

Reply:
it's the same behavior now and then, current changes encapsulate initialization of a plugin into a method, if something goes wrong an exception is thrown. That exception is then captured and logged before continuing to initialize next configured plugin if any.